### PR TITLE
fix(worker): record why a guard-rejected row settled with nothing written

### LINKED ISSUE #655: done rows settle through musixmatch with no outcome_type; 3,163 and still occurring
[mode: plan]
[model: opus]
[effort: high]

## Description

**3,163 `work_queue` rows are `status='done'`, settled through musixmatch, with an empty
`last_error` and `outcome_type` NULL.** They are clean successes that never recorded what was
written. This is happening NOW, not historically: the most recent one completed
**2026-07-25T21:33Z**, while rows completing in the same window stamp correctly.

## The invariant this violates

An **outcome** is a lyric classification: `synced`, `unsynced`, or `instrumental`. It answers
"what artifact exists for this track."

It is NOT a status. Two other states exist and are separate axes:
- **deferred** -- the work is unfinished and will retry. A benign provider miss lives here.
- **error / failure** -- something went wrong, which is distinct from a provider legitimately
  having nothing.

So a row that is `done` must carry an outcome, because `done` asserts the work reached a lyric
result. A `done` row with a NULL `outcome_type` is a contradiction: it claims settlement while
recording nothing settled.

## Evidence (production, 2026-07-25)

| property | value |
|---|---|
| rows `done` + provider lane + NULL `outcome_type` | **3,163** |
| `last_error` on those rows | **empty on all 3,163** -- not misses, not failures |
| `provider_lane` | musixmatch, all of them |
| `completed_at` range | 2026-06-20T02:30Z -> **2026-07-25T21:33Z (today)** |
| earliest row WITH a stamped outcome | 2026-06-24T00:20Z |
| all `done` rows | 34,754 |
| all `done` rows with NULL `outcome_type` | 10,937 (31.5%) -- the remaining ~7,774 carry no provider lane and are presumed genuinely legacy |

**The migration-era explanation is REFUTED.** Unstamped rows are still being produced today,
concurrently with correctly-stamped ones. This was the cheap hypothesis and it does not hold.

## Where it goes wrong

`worker.go:1290-1294` stamps the outcome just before `Complete`:

```go
if outcome := outcomeTypeFromSong(song); outcome != "" {
    if stampErr := w.queue.SetOutcomeType(ctxNoCancel, item.ID, outcome); stampErr != nil {
        slog.Warn("worker: stamp outcome type failed; continuing", "id", item.ID, "error", stampErr)
    }
}
```

Two properties of this code permit the defect:

1. **The stamp is skipped entirely when `outcomeTypeFromSong` returns `""`** -- no subtitles, no
   lyrics body, no instrumental flag. The call-site comment says that "shouldn't reach here on
   the success path," but 3,163 rows suggest otherwise.
2. **The stamp is non-fatal.** A failure logs a warning and completion proceeds, so a systematic
   stamping failure is invisible except as this NULL column.

Under the invariant above, both are wrong: stamping is part of completing a row, not a
best-effort annotation. A row that cannot be classified should not reach `done`.

## Investigation (ordered; the cheap discriminators first)

- [ ] `grep` the container logs for `stamp outcome type failed`. Non-zero proves cause (2);
      zero effectively rules it out and points at cause (1).
- [ ] Determine whether these rows have `output_paths`, and whether those files EXIST on disk.
      A `done` row with no file is a much more serious finding than a bookkeeping gap.
- [ ] Audit every path reaching `status='done'` for a missing `SetOutcomeType` -- the
      instrumental path stamps separately at `worker.go:1543`; cache hits and the settle paths
      in `internal/queue` need checking too.
- [ ] Establish what a musixmatch response looks like when `outcomeTypeFromSong` returns `""`
      but the row still completes. That is the reproduction.

## Acceptance Criteria

- [ ] The mechanism is identified with evidence, not a plausible story
- [ ] A `done` row can no longer carry a NULL `outcome_type`: either the stamp becomes part of
      completion (and fails loudly), or a row with nothing writable does not reach `done`
- [ ] Covered by a test over the reproducing case
- [ ] A decision on the existing 3,163: backfillable from `output_paths`/on-disk artifacts, or
      unknowable and marked as such
- [ ] The ~7,774 provider-less rows are separately classified as genuinely legacy, or
      backfilled

## Additional Context

Surfaced when the UI showed a scan result as a provider lane plus an outcome of `unknown`, with
no way to learn more. The reporting side of that is filed separately -- but note that fixing the
reporting alone would only hide this: the data is genuinely missing, and no UI can render a
classification that was never recorded.

### PR DESCRIPTION
**Pull request by sydlexius**
## Summary

- **A guard-rejected row settled `done` with a NULL `outcome_type`**, which reports render as "unknown" -- the same rendering as a legacy row predating the column. One value meant two unrelated things, so an operator could not tell a deliberate policy rejection from missing history. 3,182 rows sat that way on production, and the count was still growing.
- **Adds `outcome_type='rejected'`** (reported as `ResultRejected`), stamped before `Complete` on the guard path. Its own value rather than reusing `unsynced`: nothing was written, so classifying it as an artifact would be a lie reports would then count as coverage.
- **Look at first:** the mechanism was confirmed against production logs, not inferred, and both hypotheses in the issue body are refuted. See "Test plan".

## Linked issue

Closes #655

## Pre-flight checklist

- [x] Pre-push gate green locally (`make gate`), run via `/prep-pr`.
- [x] Code review pass complete; findings fixed before pushing.
  - Self-review found **two real defects in this change**, both fixed. See "Test plan".
- [x] Commits squashed into one clean commit before the first push.
- [x] Label set on `gh pr create --label ...`.
- [ ] UAT performed for user-visible changes (run the binary, not just the test suite).
  - **Not done, and not claimed.** See "Not covered".
- [x] Screenshot attached for any web-UI change -- N/A, no UI change. The new class flows through the existing reports classifier.
- [x] `make ui` re-run -- N/A, no `.templ` or CSS change.
- [x] Migration under `internal/db/migrations/` -- **deliberately none.** See "Not covered".

## Test plan

- [x] `make gate` passes locally. Exit 0, patch coverage 72.00% against a 70% threshold.
- [x] New tests were confirmed to **fail against unfixed code**.
      `TestGuardRejectedRowRecordsItsOutcome` reproduces rather than guards: removing the `SetOutcomeType` call from the guard branch reds with *"guard-rejected row settled with NO outcome_type recorded"*, at the assertion rather than the build. Restore green.
      It ships with **two controls**, because "the rejected row says rejected" is equally consistent with stamping every row that way: an accepted row must still say `synced`, and so must a row with no guard wired.
- [x] Patch coverage meets the Codecov threshold -- 72.00% (18/25).
- [x] Which **surface** this change fixes is stated explicitly and was verified by looking at it.
      The mechanism was established from production, not reasoning:
      - Guard rejections in the container logs match unstamped rows **day for day** across the entire log-retained window (13 and 13).
      - Every such row carries the exact column signature the path leaves: `outcome_type`, `fetched_at`, `writer_version` all NULL.
      - **Cause (2) refuted** -- `stamp outcome type failed` appears **zero** times across the live log and all ten archives. Verified that absence is real rather than filtered, via a control: 42 `WARN` lines present in the same log, worker-level ones among them, `level=debug`.
      - **Cause (1) refuted** -- `outcomeTypeFromSong`'s only empty arms are `Quarantine` (which stamps `timing_outcome`; **zero** unstamped rows carry one) and a `default` arm unreachable past `w.store` and `WriteLRC`.
      - `RetireMiss` was also considered and refuted: it writes `last_error`, which 3,176 of the rows lack.
- [x] Manual UAT steps:
  - [x] Read the production queue read-only (`mode=ro`) to establish the population, its column signature, and its rate over time.
  - [x] Cross-referenced container logs day-by-day against row completion dates to test the correlation rather than assume it.
  - [x] Mutation-tested the new test by removing the fix.
- [ ] Reviewer follow-ups:
  - [ ] **Is `rejected` the right model?** The alternative was a distinct `status` rather than an `outcome_type` value. That is truer to the issue's stated invariant (`done` asserts success), but `Dequeue`, the reports, and the scan writeback all key on status, so it is a much larger blast radius for the same user-visible result. I took the smaller change; say if you want the other.
  - [ ] **The June population.** 3,115 rows share the signature but predate log retention. I left them `unknown` rather than backfilling on a guess -- see "Not covered".

## Not covered

- **No migration, and no backfill of the existing 3,182 rows.** This is a deliberate answer to the issue's AC 4, not an omission. A guard rejection writes **nothing to disk**, so there is no artifact to recover a value from. The issue proposes backfilling from `output_paths`, which cannot work either: that column is marshaled at **enqueue** time (`internal/scan/enqueuer.go:288`), so its presence says nothing about what landed. All 3,182 rows have it -- as do the 29,983 correctly-stamped ones.
- **The June 20-25 bulk (3,115 rows) is consistent with this mechanism but UNVERIFIED.** Its column signature is indistinguishable from the confirmed cases, but logs do not reach back that far, and a matching signature is weaker evidence than a matching log line. Claiming it would be a guess dressed as a finding. Those rows stay `unknown`, which is the honest classification: neither the row nor the disk records why they settled.
- **The ~7,737 rows with no `provider_lane`** are untouched here. They remain the issue's AC 5 and are presumed genuinely legacy.
- **The binary was never run.** Verification was the production database, the production logs, and unit tests. Nothing here exercises a live serve-mode fetch through a configured language guard.
- **The constants are not a single point of definition**, and the comment says so rather than overclaiming. `internal/queue`'s detector settle and `internal/reports`' classifier still use SQL literals inside raw query strings, where a Go constant cannot substitute without concatenating statements. Adding a value means touching all three sites.

## Summary by CodeRabbit

* **New Features**
  * Added a distinct “Rejected” result classification for items declined by language or script safeguards.
  * Rejected items now receive a clear terminal outcome while remaining separate from failed or deferred items.
* **Bug Fixes**
  * Improved outcome reporting so rejected items are completed exactly once and processed consistently.
  * Preserved existing behavior for accepted items and workflows without safeguards.
* **Tests**
  * Added coverage for rejected, accepted, and standard processing scenarios.

### CODE REVIEW

**Comment by sydlexius** [comment]
@coderabbitai review

File: internal/worker/worker.go [hunk 1]
```diff
@@ -1369,7 +1408,26 @@ func (w *Worker) RunOnce(ctx context.Context) error {
...
+			// Stamped as its own value rather than reusing "unsynced": nothing was
+			// written here, so classifying it as an artifact would be a lie that
+			// reports would then count as coverage.
+			if stampErr := w.queue.SetOutcomeType(ctxNoCancel, item.ID, outcomeTypeRejected); stampErr != nil {
+				// Non-fatal, matching every other pre-Complete stamp: the policy
+				// decision is already correct and losing its label must not strand
+				// a row that is otherwise finished. The warning is the audit trail.
+				slog.Warn("worker: stamp guard-rejected outcome failed; continuing", "id", item.ID, "error", stampErr)
+			}
+			if err := w.queue.Complete(ctxNoCancel, item.ID); err != nil {
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply internal/worker/worker.go]: Fixed in `0b15d06`, with coverage follow-ups in `ae61785` and `9b0929f`. **You were right, and Copilot found the same thing independently — the fix could reproduce the bug it fixes.**

A best-effort stamp followed by an unconditional `Complete` means a dropped stamp plus a successful completion leaves the row `done` with a NULL `outcome_type`, which is precisely the state #655 exists to eliminate. A label that can be silently lost is not a fix for a missing label.

**Implemented as you described.** `SettleGuardRejected` writes `outcome_type='rejected'`, `status='done'` and the `scan_results` writeback in one transaction, guarded on `'processing'`. It mirrors `SettleInstrumental` deliberately — same transaction shape, same guarded UPDATE, same `SettleOutcome` vocabulary — because "record a terminal verdict and complete the row" is one operation with two verdicts, and letting the two paths diverge is how one of them drifts.

On failure the row is **not** completed: it fails and retries. The guard rejects deterministically, so a retry costs one provider round-trip and cannot loop forever on a row that would otherwise settle unlabeled.

**Two things your comment did not ask for, which the fix exposed:**

1. **A non-Settled branch.** A no-op UPDATE is not an error, so the previous code had no signal at all for a row pruned mid-flight or claimed by a concurrent writer. That case now **releases** rather than fails — failing would re-queue a row that may no longer exist. Its own error path (settle did not take *and* the release failed) is also covered in `9b0929f`, because that combination strands the row in `'processing'`, a status nothing reclaims.
2. **The queue-layer test gap.** The worker suite proves the seam is *called*; it uses a fake, so it says nothing about whether the SQL is right — the same shape of gap that let the original defect ship. `ae61785` adds real-SQLite tests for the happy path, the `processing` guard, and a missing row.

**Regression test for the persistence-failure path, as requested**, and verified rather than asserted: reverting to stamp-then-Complete reds with *"row was completed despite a failed settle"*; removing the `AND status = ?` guard reds with *"a row the caller does not hold was settled"*. Restores green in both cases.

The fake also models the transaction as **all-or-nothing**. A fake that stamped the outcome and skipped only the completion would have agreed with the buggy code and validated nothing.

File: internal/worker/worker.go [hunk 2]
```diff
@@ -1369,7 +1408,26 @@ func (w *Worker) RunOnce(ctx context.Context) error {
...
+			//
+			// Stamped as its own value rather than reusing "unsynced": nothing was
+			// written here, so classifying it as an artifact would be a lie that
+			// reports would then count as coverage.
+			if stampErr := w.queue.SetOutcomeType(ctxNoCancel, item.ID, outcomeTypeRejected); stampErr != nil {
+				// Non-fatal, matching every other pre-Complete stamp: the policy
+				// decision is already correct and losing its label must not strand
+				// a row that is otherwise finished. The warning is the audit trail.
+				slog.Warn("worker: stamp guard-rejected outcome failed; continuing", "id", item.ID, "error", stampErr)
+			}
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply internal/worker/worker.go]: Fixed in `0b15d06`. You and CodeRabbit found this independently, which is a fair signal it was the weakest thing in the diff — and you were right that the fix could regress to the exact ambiguity it was meant to remove.

Rather than making the stamp fatal alongside a separate `Complete`, the two are now **one transaction**: `SettleGuardRejected` writes `outcome_type='rejected'`, `status='done'` and the `scan_results` writeback together, guarded on `'processing'`, or nothing lands. That is stronger than a fatal stamp — a fatal stamp still leaves a window where the stamp commits and the completion does not — and it matches the existing `SettleInstrumental` precedent rather than inventing a second shape for the same operation.

Your "or otherwise preventing completion" framing is what the failure path now does: on error the row is **not** completed, it fails and retries. The guard rejects deterministically, so a retry costs one provider round-trip and cannot spin on a row that would otherwise settle unlabeled.

Verified rather than asserted: reverting to stamp-then-Complete reds with *"row was completed despite a failed settle"*, and the restore is green.

File: internal/reports/reports.go [hunk 3]
```diff
@@ -139,6 +161,7 @@ func (r *Repo) RecentOutcomes(ctx context.Context, limit int) ([]RecentOutcome,
                 WHEN outcome_type = 'synced' THEN 'synced'
                 WHEN outcome_type = 'unsynced' THEN 'unsynced'
                 WHEN outcome_type = 'instrumental' THEN 'instrumental'
+                WHEN outcome_type = 'rejected' THEN 'rejected'
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply internal/reports/reports.go:167]: Fixed in `0b15d06`.

Correct, and the reasoning is the important part: the worker test proves the value is *persisted*, but it cannot see the SQL `CASE` in `RecentOutcomes`. A change there would silently fall back to `unknown` -- which is the original #655 symptom, not a new one.

`TestRecentOutcomesClassificationAndOrder` now carries a `rejected` fixture asserting `ResultRejected`, alongside the existing synced/unsynced/instrumental/miss/legacy rows. It sits in the same ordering table, so it also covers the sort position rather than only the mapping.

One detail worth flagging for anyone reading the diff: adding a row shifted the positional assertions further down the test (`got[3]`/`got[4]` for the field carry-through and the NULL `completed_at` check). Those were updated rather than left to pass by coincidence.

File: internal/reports/reports.go [hunk 4]
```diff
@@ -139,6 +161,7 @@ func (r *Repo) RecentOutcomes(ctx context.Context, limit int) ([]RecentOutcome,
...
                 ELSE 'unknown'
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply internal/reports/reports.go:168]: Fixed in `0b15d06`. You and CodeRabbit raised this independently.

`TestRecentOutcomesClassificationAndOrder` now carries a `rejected` fixture asserting `ResultRejected`, in the same ordering table as the existing synced/unsynced/instrumental/miss/legacy rows -- so it covers the sort position as well as the mapping.

Your framing of the risk is exactly the one that matters: the worker test proves the value is *persisted*, but nothing could see the SQL `CASE`. A change there would silently fall back to `unknown`, which is not a new bug -- it is #655's original symptom returning by a different route.

Adding the row shifted the positional assertions later in the test (`got[3]`/`got[4]` for field carry-through and the NULL `completed_at` check); those were updated rather than left to pass by coincidence.

File: internal/reports/reports.go [hunk 5]
```diff
@@ -124,10 +140,16 @@ type RecentOutcome struct {
 // Source: work_queue rows where status='done'. The result classification is
 // computed in SQL from the recorded outcome_type (stamped at completion, #379),
 // NOT the output_paths filename: last_error='miss limit reached' -> miss;
-// otherwise outcome_type 'synced'/'unsynced'/'instrumental' map to the matching
-// ResultClass; a NULL outcome_type (a legacy row predating the column) -> unknown.
+// otherwise outcome_type 'synced'/'unsynced'/'instrumental'/'rejected' map to
+// the matching ResultClass; a NULL outcome_type -> unknown.
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply internal/reports/reports.go:147]: Fixed in `0b15d06`.

Correct -- and this was the third copy of the same stale claim in one file. `RecentOutcomes`' own doc comment and the `ResultUnknown` constant both said classification came from `output_paths` or that a NULL `outcome_type` meant a legacy row; I corrected those two during self-review and missed the struct field.

That is worth noting rather than just fixing: three statements of one fact is how one of them goes stale, and in this case the stale statement was itself #655. The claim that NULL meant "legacy" is precisely what made a live, ongoing policy rejection indistinguishable from missing history.

The field doc now says the classification comes from `last_error` / `outcome_type`, with the reason `output_paths` cannot serve: it holds the enqueue-time `.lrc` plan and is never updated at completion, which is what made every completed row read as synced before #379.

**Comment by sydlexius** [comment]
@coderabbitai approve
